# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: widget/windows/TSFTextStore.cpp
# Commit: 46bf06a42ba6
# Full Hash: 46bf06a42ba623133cfdaceac9d3f9f778975b79
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2020-03-26 15:34:27
# Description:
#   Bug 1589497 - Make PendingAction of TSFStore copy of key message r=m_kato
#   
#   `TSFTextStore::sHandlingKeyMsg` refers pointer of struct, but referred via
#   `TSFTextStore::PendingAction` so that we should make it has a copy of
#   `sHandlingKeyMsg` because of for async handling.
# ==============================================================================

diff -r 1fe283e3bb59 -r 46bf06a42ba6 widget/windows/TSFTextStore.cpp
--- a/widget/windows/TSFTextStore.cpp	Thu Mar 26 07:18:23 2020 +0000
+++ b/widget/windows/TSFTextStore.cpp	Thu Mar 26 07:33:37 2020 +0000
@@ -2341,15 +2341,15 @@
     switch (action.mType) {
       case PendingAction::Type::eKeyboardEvent:
         if (mDestroyed) {
-          MOZ_LOG(
-              sTextStoreLog, LogLevel::Warning,
-              ("0x%p   TSFTextStore::FlushPendingActions() "
-               "IGNORED pending KeyboardEvent(%s) due to already destroyed",
-               action.mKeyMsg->message == WM_KEYDOWN ? "eKeyDown" : "eKeyUp",
-               this));
+          MOZ_LOG(sTextStoreLog, LogLevel::Warning,
+                  ("0x%p   TSFTextStore::FlushPendingActions() "
+                   "IGNORED pending KeyboardEvent(%s) due to already destroyed",
+                   action.mKeyMsg.message == WM_KEYDOWN ? "eKeyDown" : "eKeyUp",
+                   this));
         }
-        MOZ_DIAGNOSTIC_ASSERT(action.mKeyMsg);
-        DispatchKeyboardEventAsProcessedByIME(*action.mKeyMsg);
+        MOZ_DIAGNOSTIC_ASSERT(action.mKeyMsg.message == WM_KEYDOWN ||
+                              action.mKeyMsg.message == WM_KEYUP);
+        DispatchKeyboardEventAsProcessedByIME(action.mKeyMsg);
         if (!widget || widget->Destroyed()) {
           break;
         }
@@ -2665,7 +2665,7 @@
          this));
     PendingAction* action = mPendingActions.AppendElement();
     action->mType = PendingAction::Type::eKeyboardEvent;
-    action->mKeyMsg = sHandlingKeyMsg;
+    memcpy(&action->mKeyMsg, sHandlingKeyMsg, sizeof(MSG));
     return;
   }
 